File: php/play.php
<?php

error_reporting(E_ALL);

require_once 'world.php';

final class Play {
  private const int WORLD_WIDTH = 150;
  private const int WORLD_HEIGHT = 40;

  public static function run(): void {
    $world = new World(
      width: self::WORLD_WIDTH,
      height: self::WORLD_HEIGHT,
    );

    $minimal = getenv("MINIMAL") !== false;

    if (!$minimal) {
      echo $world->render();
    }

    $totalTick = 0;
    $lowestTick = INF;
    $totalRender = 0;
    $lowestRender = INF;

    while (true) {
      $tickStart = hrtime(true);
      $world->doTick();
      $tickFinish = hrtime(true);
      $tickTime = $tickFinish - $tickStart;
      $totalTick += $tickTime;
      $lowestTick = min($lowestTick, $tickTime);
      $avgTick = $totalTick / $world->tick;

      $renderStart = hrtime(true);
      $rendered = $world->render();
      $renderFinish = hrtime(true);
      $renderTime = $renderFinish - $renderStart;
      $totalRender += $renderTime;
      $lowestRender = min($lowestRender, $renderTime);
      $avgRender = $totalRender / $world->tick;

      if (!$minimal) {
        echo "\u{001b}[H\u{001b}[2J";
      }

      printf(
        "#%d - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)\n",
        $world->tick,
        self::_f($lowestTick),
        self::_f($avgTick),
        self::_f($lowestRender),
        self::_f($avgRender)
      );

      if (!$minimal) {
        echo $rendered;
      }
    }
  }

  private static function _f(int|float $value): float {
    // nanoseconds -> milliseconds
    return $value / 1_000_000;
  }
}

Play::run();

File: nim/play.nim
import std/[envvars, monotimes, times, strutils, strformat]
include world

const WorldWidth = 150
const WorldHeight = 40
const ClearScreen = "\x1b[?2026h\x1b[H\x1b[2J"
const ShowScreen = "\x1b[?2026l"

type
  Play = ref object

# By default, Nim requires methods be declared before they are used elsewhere
# and will error out if I dont. To to order the methods as I like, I need to
# declare them ahead of time, known as "forward declaration".
proc run(self: Play)
func f(self: Play, value: float): float

proc run(self: Play) =
  let world = World(
    width: WorldWidth,
    height: WorldHeight,
  ).initialize()

  let minimal = getEnv("MINIMAL") != ""

  if not minimal:
    echo world.render()

  var totalTick = 0.0
  var lowestTick = Inf
  var totalRender = 0.0
  var lowestRender = Inf

  while true:
    let tickStart = getMonoTime()
    world.doTick()
    let tickFinish = getMonoTime()
    let tickTime = (tickFinish - tickStart).inNanoseconds.float
    totalTick += tickTime
    lowestTick = min(lowestTick, tickTime)
    let avgTick = (totalTick / world.tick.float)

    let renderStart = getMonoTime()
    let rendered = world.render()
    let renderFinish = getMonoTime()
    let renderTime = (renderFinish - renderStart).inNanoseconds.float
    totalRender += renderTime
    lowestRender = min(lowestRender, renderTime)
    let avgRender = (totalRender / world.tick.float)

    if not minimal:
      stdout.write ClearScreen

    echo fmt"""
      #{world.tick} -
      World Tick (L: {self.f(lowestTick):.3f}; A: {self.f(avgTick):.3f}) -
      Rendering (L: {self.f(lowestRender):.3f}; A: {self.f(avgRender):.3f})
    """.dedent().replace("\n", " ")

    if not minimal:
      echo rendered & ShowScreen

func f(self: Play, value: float): float =
  # nanoseconds -> milliseconds
  value / 1_000_000

Play().run()
